---
title: "VariantsInvestigator"
output: 
  flexdashboard::flex_dashboard:
    orientation: column
    vertical_layout: fill
runtime: shiny
---

```{r global, include=FALSE}
library(flexdashboard)
library(tibble)
library(dbplyr)
library(dplyr)
library(RSQLite)
library(DBI)
library(tidyr)
library(flexdashboard)
library(shiny)
library(purrr)
library(janitor)
library(gt)


con <- DBI::dbConnect(RSQLite::SQLite(), dbname = "test_variants_db.sqlit")
gene_name <- function(name){variants_db %>%
  filter(SYMBOL_1 %like% name)}
variants_db <- tbl(con, "variants")
polyphen <- function(minpol, maxpol){variants_db %>%
    filter(between(PolyPhen_values, minpol, maxpol))
}

giant_filter <- function(minsft, maxsft, minpol, maxpol, gnomin, gnomax){variants_db %>%
    dplyr::filter(between(SIFT_values, minsft, maxsft) & 
                    between(PolyPhen_values, minpol, maxpol) &
                    between(gnomAD_AF_1, gnomin, gnomax))
}
freq_gnomAD <- function(gnomin, gnomax){dplyr::filter(between(gnomAD_AF_1, gnomin, gnomax))
}

flag_lof <- function(){variants_db %>%
    filter(LoF_1 == "HC")}
```

Column {.sidebar}
-----------------------------------------------------------------------

### Filtering control widgets

```{r}
library(shinyWidgets)

sliderInput(inputId = "gnomAD", "Maximum gnomAD allele frequency", min = 0, max = 1,
                       step = 0.001, value = 1)
sliderInput(inputId = "geno_qual", "minimum genotype quality", min = 0, max = 99,
                       step = 1, value = 0)
sliderInput(inputId = "read_dp", "min/max read depth", min = 0, max = 200,
                       step = 1, value = c(0,200))
sliderInput(inputId = "SIFT", "Maximum SIFT value", min = 0, max = 1,
                       step = 0.01, value = 1)
sliderInput(inputId = "PolyPhen", "Minimum PolyPhen value", min = 0, max = 1,
                       step = 0.01, value = 0)

pickerInput(inputId = "consh", 
            label = "high impact consequences", 
            choices = list(
                           "transcript_ablation" = "transcript_ablation",
                           "splice_acceptor_variant" = "splice_acceptor_variant",
                           "splice_donor_variant" = "splice_donor_variant",
                           "stop_gained" = "stop_gained",
                           "frameshift_variant" = "frameshift_variant",
                           "stop_lost" = "stop_lost",
                           "start_lost" = "start_lost",
                           "transcript_amplification" = "transcript_amplification",
                           "feature_elongation" = "feature_elongation",
                           "feature_truncation" = "feature_truncation"
                           ), 
            options = list(`actions-box` = TRUE), 
            multiple = TRUE)

pickerInput(inputId = "consmod", 
            label = "moderate impact consequences", 
            choices = list("inframe_insertion" = "inframe_insertion",
                           "inframe_deletion" = "inframe_deletion",
                           "missense_variant" = "missense_variant",
                           "protein_altering_variant" = "protein_altering_variant"
                           ), 
            options = list(`actions-box` = TRUE),
            multiple = TRUE)

pickerInput(inputId = "consl", 
            label = "low impact consequences", 
            choices = list("splice_donor_5th_base_variant" = "splice_donor_5th_base_variant",
                           "splice_region_variant" = "splice_region_variant",
                           "splice_donor_region_variant" = "splice_donor_region_variant",
                           "splice_polypyrimidine_tract_variant" = "splice_polypyrimidine_tract_variant",
                           "incomplete_terminal_codon_variant" = "incomplete_terminal_codon_variant",
                           "start_retained_variant" = "start_retained_variant",
                           "stop_retained_variant" = "stop_retained_variant",
                           "synonymous_variant" = "synonymous_variant"), 
            options = list(`actions-box` = TRUE),
            multiple = TRUE)

pickerInput(inputId = "consmdf", 
            label = "modifier impact consequences", 
            choices = list(
                           "coding_sequence_variant" = "coding_sequence_variant",
                           "mature_miRNA_variant" = "mature_miRNA_variant",
                           "5_prime_UTR_variant" = "5_prime_UTR_variant",
                           "3_prime_UTR_variant" = "3_prime_UTR_variant",
                           "non_coding_transcript_exon_variant" = "non_coding_transcript_exon_variant",
                           "intron_variant" = "intron_variant",
                           "NMD_transcript_variant" = "NMD_transcript_variant",
                           "non_coding_transcript_variant" = "non_coding_transcript_variant",
                           "coding_transcript_variant" = "coding_transcript_variant",
                           "upstream_gene_variant" = "upstream_gene_variant",
                           "downstream_gene_variant" = "downstream_gene_variant",
                           "TFBS_ablation" = "TFBS_ablation",
                           "TFBS_amplification" = "TFBS_amplification",
                           "TF_binding_site_variant" = "TF_binding_site_variant",
                           "regulatory_region_ablation" = "regulatory_region_ablation",
                           "regulatory_region_amplification" = "regulatory_region_amplification",
                           "regulatory_region_variant" = "regulatory_region_variant",
                           "intergenic_variant" = "intergenic_variant",
                           "sequence_variant" = "sequence_variant"
                           ),
            options = list(`actions-box` = TRUE),
            multiple = TRUE
            )
  
checkboxGroupInput("loft", 
                   "Loftee High Confidence", 
                   choices = list("HC" = 1)
                   )


textInput(inputId = "gene_symbol", "gene symbol")

checkboxInput("submitexon", label="exonic variant")
textInput(inputId = "EXON", "specific exon")
checkboxInput("submitintron", label="intronic variant")
textInput(inputId = "INTRON", "specific intron")

checkboxGroupInput("LoF", "Loftee High Confidence", choices = list("HC" = 1))







  




```
Row {data-width=800}
-----------------------------------------------------------------------
### Chart D
```{r}
giant_filter <- function(minsft, maxpol, gnomax, gq, consh, consmod, consl, consmdf){
   variants_db %>%
    dplyr::filter(
      SIFT_values <= minsft | is.na(SIFT_values)==TRUE & 
      PolyPhen_values >= maxpol | is.na(PolyPhen_values)==TRUE &
      gnomAD_AF_1 <= gnomax | is.na(gnomAD_AF_1)==TRUE &
      GQ >= gq &
      filter(Consequence %in% consh) & 
        filter(Consequence %in% consmod) & filter(Consequence %in% consl) & filter(Consequence %in% consmdf)
      )
}



render_gt({giant_filter(
                        input$SIFT,
                        input$polyphen,
                        input$gnomAD,
                        input$geno_qual,
                        input$consh,
                        input$consmod,
                        input$consl,
                        input$consmdf) %>%
    dplyr::select("start","end", "REF", "ALT", "PolyPhen_values", "PolyPhen_prediction", "SIFT_values", "SIFT_prediction", "gnomAD_AF_1", "Consequence", "IMPACT_1") %>%
    gt()%>%
    opt_interactive(
    use_search = TRUE,
    use_filters = TRUE,
    use_resizers = TRUE,
    use_highlight = TRUE,
    use_compact_mode = TRUE,
    use_text_wrapping = FALSE,
    use_page_size_select = TRUE
  ) %>%
    data_color(
    columns = PolyPhen_prediction,
    palette = "Blues"
  )%>%
    data_color(
    columns = SIFT_prediction,
    palette = "Greens"
  )})
 
```
Column {data-width=150}
-----------------------------------------------------------------------

### Chart B

```{html}

```

### Chart C

```{r}

```

